File: src/main/java/io/dev/mutex/thesisinfomgmt/service/ThesisService.java
package io.dev.mutex.thesisinfomgmt.service;

import io.dev.mutex.thesisinfomgmt.common.PaginatedData;
import io.dev.mutex.thesisinfomgmt.dto.ThesisDTO;

public interface ThesisService {

  /**
   * Retrieve theses
   *
   * @param page     the page of the data to be retrieved
   * @param pageSize the number of items to be retrieved
   * @param query    the query string
   * @return paginated thesis data
   */
  PaginatedData<ThesisDTO> getTheses(int page, int pageSize, String query);

  /**
   * Creates a new thesis
   *
   * @param thesis the details of the new thesis
   * @return created thesis
   */
  ThesisDTO createThesis(ThesisDTO thesis);

  /**
   * Retrieves a thesis by its id
   *
   * @param id the id of the thesis to be retrieved
   * @return retrieved thesis
   */
  ThesisDTO getThesis(long id);

  /**
   * Deletes a thesis
   *
   * @param id the id of the thesis to be deleted
   */
  void deleteThesis(long id);

  /**
   * Updates a thesis
   *
   * @param id     the id of the thesis to be updated
   * @param thesis the updated thesis details
   */
  void updateThesis(long id, ThesisDTO thesis);
}

File: app/src/services/ThesisService.ts
import { PaginatedData } from "../types/Common";
import { Thesis } from "../types/Thesis";
import RequestHelper from "../helpers/RequestHelper";
import qs from "qs";
import { RequestMethod } from "../common/constants";

class ThesisService {
  resourcePath = "theses";

  async getTheses(
    page: number,
    pageSize: number,
    query: string
  ): Promise<PaginatedData<Thesis>> {
    const queryParams = {
      page: page,
      pageSize: pageSize,
      query: query,
    };
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}?${qs.stringify(queryParams)}`,
      "get theses"
    );
  }

  async createThesis(thesis: Thesis): Promise<Thesis> {
    return await RequestHelper.doRemoteCallAsync(
      this.resourcePath,
      "create thesis",
      RequestMethod.POST,
      thesis
    );
  }

  async updateThesis(thesisId: number, thesis: Thesis): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${thesisId}`,
      "update thesis",
      RequestMethod.PUT,
      thesis
    );
  }

  async deleteThesis(thesisId: number): Promise<void> {
    return await RequestHelper.doRemoteCallAsync(
      `${this.resourcePath}/${thesisId}`,
      "delete thesis",
      RequestMethod.DELETE
    );
  }
}

export default new ThesisService();
